messaging/controller/processor/processor.go

```diff
@@ -0,0 +1,335 @@
+	"github.com/status-im/status-go/pkg/pubsub"
+)
+
+const (
+	maxNumOfEphemeralKeys = 3
+)
+
+var errReliabilityNotStarted = errors.New("reliability not started")
+
+type Processor struct {
+	identity *ecdsa.PrivateKey
+	stack    *common.MessagingStack
+
+	messageConfirmationStorage common.MessageConfirmationPersistence
+	hashRatchetStorage         common.HashRatchetPersistence
+
+	ephemeralKeysManager *EphemeralKeysManager
+
+	publisher *pubsub.Publisher
+	logger    *zap.Logger
+}
+
+func NewProcessor(
+	identity *ecdsa.PrivateKey,
+	stack *common.MessagingStack,
+	messageConfirmationStorage common.MessageConfirmationPersistence,
+	hashRatchetStorage common.HashRatchetPersistence,
+	logger *zap.Logger,
+) *Processor {
+	return &Processor{
+		identity:                   identity,
+		stack:                      stack,
+		messageConfirmationStorage: messageConfirmationStorage,
+		hashRatchetStorage:         hashRatchetStorage,
+		ephemeralKeysManager:       NewEphemeralKeysManager(maxNumOfEphemeralKeys),
+		publisher:                  pubsub.NewPublisher(),
+		logger:                     logger.Named("receiver"),
+	}
+}
+
+func (r *Processor) Publisher() *pubsub.Publisher {
+	return r.publisher
+}
+
+func (r *Processor) GetEphemeralKey() (*ecdsa.PrivateKey, error) {
+	key, err := r.ephemeralKeysManager.GenerateOrGetRandom()
+	if err != nil {
+		return nil, err
+	}
+	_, err = r.stack.Transport.LoadKeyFilters(key)
+	if err != nil {
+		return nil, err
+	}
+	return key, nil
+}
+
+func (r *Processor) ProcessMessage(msg *types.ReceivedMessage) (*types.HandleMessageResponse, error) {
+	response, err := r.processMessage(msg)
+	if response == nil || err != nil {
+		return nil, err
+	}
+
+	// Process queued hash ratchet messages
+	for _, message := range response.messages {
+		queuedMessagesResponse, err := r.processQueuedHashRatchetMessages(message.EncryptionLayer.HashRatchetInfo)
+		if err != nil {
+			return nil, err
+		}
+		response.messages = append(response.messages, queuedMessagesResponse.messages...)
+		response.ackedMessageIDs = append(response.ackedMessageIDs, queuedMessagesResponse.ackedMessageIDs...)
+	}
+
+	return &types.HandleMessageResponse{
+		Messages:        response.messages,
+		AckedMessageIDs: response.ackedMessageIDs,
+	}, nil
+}
+
+type processMessageResponse struct {
+	messages        []*types.Message
+	ackedMessageIDs []cryptotypes.HexBytes
+}
+
+func (r *Processor) processMessage(m *types.ReceivedMessage) (*processMessageResponse, error) {
+	logger := r.logger.With(zap.Stringer("hash", cryptotypes.HexBytes(m.Hash)))
+	logger.Debug("processing received message")
+
+	responseMessage := &types.Message{}
+
+	response := &processMessageResponse{
+		messages:        []*types.Message{responseMessage},
+		ackedMessageIDs: []cryptotypes.HexBytes{},
+	}
+
+	err := processTransportLayer(responseMessage, m)
+	if err != nil {
+		logger.Error("failed to process transport layer", zap.Error(err))
+		return nil, err
```

Comment:
Something I'm doing in implementing a `StatusBackend` class:
```suggestion
		return nil, errors.Wrap(err, "failed to process transport layer")
```

I think it's good to **always** wrap the returned errors. Otherwise it's a pain in the ass to track later.